Source/ProgressionSystemRuntime/Private/Data/PSWorldSubsystem.cpp

```diff
@@ -255,87 +262,53 @@ void UPSWorldSubsystem::AddProgressionStarActors()
+		UStaticMeshComponent* MeshComponent = SpawnedActor.FindComponentByClass<UStaticMeshComponent>();
+
+		SpawnedStarActorsInternal.Add(&SpawnedActor);
+
+		SpawnedActor.SetActorTransform(CurrentRowData.StarActorTransform);
+
+		// if the actor is the first element it set initial position
+		// from the initial position there is a distance between stars 
+		if (SpawnedStarActorsInternal.Num() > 1)
 		{
-			fractionalPart = modff(CurrentAmountOfUnlocked, &integerPart);
-			// check if it has a fractional part and if there is no more fully achieved stars 
-			if (fractionalPart < 1.0f && CurrentAmountOfUnlocked < 1.0f)
-			{
-				UpdateStarActor(CreatedObject, fractionalPart, 0);
-			}
-			else
-			{
-				UpdateStarActor(CreatedObject, 1, 0);
-			}
-
-			CurrentAmountOfUnlocked--;
-			continue;
+			SpawnedActor.SetActorLocation(SpawnedStarActorsInternal[SpawnedStarActorsInternal.Num() - 2]->GetActorLocation() + CurrentRowData.OffsetBetweenStarActors);
 		}
 
-		if (CurrentAmountOfLocked > 0)
+		float StarAmount = FMath::Clamp(AmountUnlocked, 0.0f, 1.0f);
+		if (AmountUnlocked > 0)
+		{
+			SetOrUpdateStarActorMesh(MeshComponent, StarAmount, false);
+		}
+		else
 		{
-			UpdateStarActor(CreatedObject, 0, 1);
-			CurrentAmountOfLocked--;
+			SetOrUpdateStarActorMesh(MeshComponent, 1, true);
 		}
+		AmountUnlocked -= StarAmount;
 	}
 }
 
 // Updates star actor to locked/unlocked according to input amounnt
-void UPSWorldSubsystem::UpdateStarActor(const FPoolObjectData& CreatedData, float AmountOfUnlockedStars, float AmountOfLockedStars)
+void UPSWorldSubsystem::SetOrUpdateStarActorMesh(UStaticMeshComponent* MeshComponent, float AmountOfStars, bool bIsLockedStar)
 {
-	AActor& SpawnedActor = CreatedData.GetChecked<AActor>();
-	UStaticMeshComponent* MeshComponent = SpawnedActor.FindComponentByClass<UStaticMeshComponent>();
-
-	const FPSRowData& CurrentRowData = GetCurrentRow();
-
-	SpawnedStarActorsInternal.Add(&SpawnedActor);
-
-	SpawnedActor.SetActorTransform(CurrentRowData.StarActorTransform);
-	// if the actor is the first element it set initial position
-	// from the initial position there is a distance between stars 
-	if (SpawnedStarActorsInternal.Num() > 1)
-	{
-		SpawnedActor.SetActorLocation(SpawnedStarActorsInternal[SpawnedStarActorsInternal.Num() - 2]->GetActorLocation() + CurrentRowData.OffsetBetweenStarActors);
-	}
-
-	if (AmountOfUnlockedStars > 0) //unlocked stars
+	if (!bIsLockedStar) // unlocked stars
 	{
-		if (AmountOfUnlockedStars > 0 && AmountOfUnlockedStars < 1) // dynamic 
+		if (AmountOfStars > 0 && AmountOfStars < 1) // stars with fractional number (e.g. 0.5) 
 		{
 			MeshComponent->SetMaterial(0, StarDynamicProgressMaterial);
-			StarDynamicProgressMaterial->SetScalarParameterValue(TEXT("Percentage2"), AmountOfUnlockedStars / 3);
+			StarDynamicProgressMaterial->SetScalarParameterValue(TEXT("Percentage2"), AmountOfStars / 3); // 3 is hardcoded value to tweak bad UV to look like it's working
```

Comment:
Hardcoding values like "Percentage2" and 3 is considered bad practice in programming and not friendly for content developers.

Both values `Percentage2` and `3` should be expanded to the data asset, like 
```
UPROPERTY(...)
FName StarMaterialSlotDivisor = TEXT("Percentage2");

UPROPERTY(...)
float StarMaterialFractionalDivisor = 3.f;
```

Reply:
done, https://github.com/h4rdmol/ProgressionSystem/pull/17/commits/ffda5ee3559c66f672d9efc79b3880c68afce272. 
Also added mentioned that 3 is temporary, till UV will be fixed